### ISSUE TITLE
PlotWrapper needs a lot of improvement or needs to be eliminated

### ISSUE THREAD

**Issue by jlstevens** [ioam/holoviews#168]
The idea of ``PlotWrapper`` is to allow either ``GridPlot`` or ``RasterGridPlot`` to be selected according to the nature of the grid space to be displayed. Unfortunately, on close inspection this class is a hack that manages to keep the tests passing but is otherwise a fairly broken concept otherwise.

The basic issue is that it doesn't work with the documentation system at all. Firstly it needs a ``renderer`` attribute (easy fix) but it also needs to pretend to be a parameterized object which leads to the following problems:

* The help system is looking for the ``get_param_values()`` method which doesn't exist.  This shouldn't be happening at all because it is supposed to be getting parameter values from the *class* but is instead looking for parameters on a parameterized *instance*. This is because ``PlotWrapper`` inherits from ``object`` and not from ``param`` causing a lot of confusion.

As the plot system is supposed to work with instances as well of classes, ``get_param_values()`` would need to be implemented anyway, and this is not possible right now because of the  following problems.

*  The parameters of ``GridPlot`` are a strict super-set of those of ``RasterGridPlot``. If ``PlotWrapper`` pretends to be the former, every option not supported by ``RasterGridPlot`` will be ignored, even though the documentation will pretend that those parameters are supported. This is an issue whether you are trying to document instances or classes.

* The parameter values are different for the parameters that *are* shared. E.g ``aspect= 1.25`` for ``RasterGridPlot`` but is set to 'auto' for ``GridPlot``. I've notices that ``fig_inches`` is different and there may be more discrepancies I haven't spotted. Again, this is a problem at both the instance and class level.

The idea of wrapping multiple classes and switching between them while presenting a fake interface of a parameterized class/object isn't particularly great to begin with.  It could be tolerable *if* the wrapped plot classes actually do *have* identical interfaces and if ``PlotWrapper`` explicitly enforces this.

As it is, the pseudo-plot the wrapper doesn't work with the documentation system, and even if it did, it would have to lie about the actual interface and parameters of wrapped contents, leading to misleading user documentation when using ``hv.help``. I don't think we are happy to let that stand!

Here are the four solutions I can think of:

1. Make ``RasterGridPlot`` and ``GridPlot`` actually support the same interface with the same default parameter values as well as the same style options. Then the difference between them comes down to optimizing special cases (grid plots containing only rasters).

2. Check for ``PlotWrapper`` everywhere to get the actual applicable plot class and then use that for documentation. This would be painful and messy as it would lead to lots of conditional logic in lots of places.

3. Implement ``GridPlot`` as a proper plotting class that makes use of ``GenericGridPlot`` and ``RasterGridPlot`` appropriately.

4. Improve the interface to ``Store.registry`` to make it more flexible when looking up a plot. Right now it is a nested dictionary of backends, the ``element:plot`` associations. Looking at the source, we access the registry a lot of times but only populate it in ``Store.register``.

  Currently, you would access a plot class using something like ```Store.registry['matplotlib'][Image]``` but   maybe ``registry`` this should be a method instead i.e ```Store.registry('matplotlib', Image)`` which returns the plot class. Then when you supply associations to ``Store.register`` your associations dictionary can include ``element:function`` items (where function takes the object and returns the plot class). Now by checking whether the value in ``Store._registry`` is a HoloViews object or not, the ``Store.registry`` method can resolve the plot class correctly via the function.

I strongly dislike option 2. If option 1. is feasible, then I don't mind keeping ``PlotWrapper`` - at least it won't be lying in the documentation about the plot/style options. Option 3. is one correct approach but I can understand why Philipp would be reluctant to implement this (might be a lot of work).

I think under the circumstances, option 4 is the clear winner as it gets rid of ``PlotWrapper`` entirely without forcing us to exactly match the interfaces of ``RasterGridPlot`` and ``GridPlot``. It would also make sense as part of the refactor of #141 and finish of the generalized rendering system without relying on hacks that end up generating misleading documentation.

Thoughts?

**Edit 1**: Unfortunately even option 4 won't help if you supply a class and not an instance (unless you also implement option 3 as well):

```python
hv.help(hv.GridSpace, visualization=True)
```

In this case, there is no right answer as to which particular plotting class should be documented. Maybe you have to something a little ridiculous and show the documentation for *all* the possible associated plotting classes?

**Edit 2**: Maybe the documentation system could reply with something like:

   *GridSpace is associated with plotting classes: GridPlot, RasterGridPlot.*

   *For help, either use ``hv.help(GridPlot)`` or ``hv.help(RasterGridPlot)``*

**Comment by jlstevens** [comment]
I think option 4 together with my suggestion in **Edit 2** would work.  It does imply a few more (relatively minor) changes:

* Support for documenting ``Plot`` classes in the same way as objects with ``visualization=True`` (should be fairly trivial).

* Not supplying any old function in the associations dictionary to ``Store.register`` but a parameterized function with a ``plots`` attribute (so the help system can find out which are the potential plot classes that may be used)

**Comment by jbednar** [comment]
And I guess just merging the two grid plot classes into a single class whose behavior changes depending on the contained objects isn't feasible?

**Comment by jlstevens** [comment]
After spending a while discussing this with Philipp, we've come up with a plan:

* In the end, we think ``RasterGridPlot`` can just be an optimized version of ``GridPlot`` with the same style options and plot options (suggestion 1 above).

* Keep ``PlotWrapper`` but rename it to ``PlotSelector`` and explicitly enforce that all the allowable plots have the same plot and style options.

* Give ``PlotSelector`` a ``plot_classes`` attribute which is an ``OrderedDict`` of name keys to plot class values. ``PlotSelector`` requires a function that given an instance returns the appropriate key.

* Check for ``PlotSelector`` instances in the help system and in ``Store.register`` and use the first plot class in the ``plot_classes`` ordered dictionary to lookup the style/plot options. All the ``plot_classes`` are looked up in ``Store.add_style_opts`` to add the specified options to all the plot classes in the selector.

* Keep ``Store.registry`` but add ``Store.plot_class`` method which behaves like the proposed ``registry`` method above (not strictly necessary to make this all work but we think it will remove some redundant code).

I think that will keep everything working sensibly without causing unnecessary confusion in the documentation. Here ``PlotSelector`` is used to switch to a more optimized plotting class when necessary but in future, there may be other reasons to dynamically switch between plot classes supporting identical interfaces.

**Comment by jbednar** [comment]
Sounds good.

**Comment by jlstevens** [comment]
Unfortunately, looking at the parameters, ``RasterGridPlot`` and ``GridPlot`` *are* different because ``RasterGridPlot`` inherits from ``OverlayPlot``.

The only thing to do now is to disable all the extra parameters of ``RasterGridPlot`` by setting a negative precedence. Then ``PlotSelector`` can ignore parameters with negative precedence when checking that the interfaces match exactly.

The good news is that these extra options were not exposed or usable *anyway* so there are no backward compatibility issues to worry about. In effect ``RasterGridPlot``'s extra options were already shadowed by the common subset defined in ``GridPlot``.

In future, Philipp may be able to go through and re-enable plot options on *both* ``GridPlot`` and ``RasterGridPlot``. The parameters of interest that may be re-enabled are: ``xticks, yticks, bgcolor, xaxis, yaxis, show_grid, invert_xaxis, invert_yaxis``.

**Comment by jlstevens** [comment]
I've reimplemented ``PlotWrapper`` as ``PlotSelector`` in commit 0f3c4db in a way that I think is much more principled and less likely to cause confusion/trouble in the long run.

In the end I opted to leave ``Store.registry`` alone and also fake just enough of the param interface to get it working again, just to minimize the need for conditional logic elsewhere.

Anyway, tests are passing and I think I'll close this issue now.
